### PR TITLE
Implement fold, foldMap and toList for the impure tree

### PR DESCRIPTION
**Pull request by hverr**
- [x] includes tests
- [x] ready for review
- [x] reviewed by @skeuchel 

#### What does this PR do?
 - Implement fold, foldMap and toList for the impure tree (+ tests)
 - Reorganize the test properties into separate files

#### Where should the reviewer start?
  1. Review `Data.BTree.Fold`
  2. Review the `readTransact` funciton in `Data.BTree.Alloc.Append`
  2. Review `tests/Properties/Fold.hs`

### CODE REVIEW

File: src/Data/BTree/Fold.hs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      => (a -> b -> b) -> b -> Tree k a -> m b
+foldr _ x (Tree _ Nothing) = return x
+foldr f x (Tree h (Just nid)) = foldrId (liftM2 f) (return x) h nid
+
+foldrId :: (AllocM m, Key k, Value a)
+        => (m a -> m b -> m b) -> m b -> Height h -> NodeId h k a -> m b
+foldrId f x h nid = readNode h nid >>= foldrNode f x h
+
+foldrNode :: (AllocM m, Key k, Value a)
+          => (m a -> m b -> m b) -> m b -> Height h -> Node h k a -> m b
```
Comment (skeuchel) [inline src/Data/BTree/Fold.hs]: Usually you fold with a algebraic list algebra that has the type (a -> b -> m b, b). Your type is not algebraic. I don't see a real problem, but I think it's still better to go with the convention. Can you implement
foldr :: (AllocM m, Key k, Value a)
      => (a -> b -> b) -> b -> Tree k a -> m b
foldr f = foldrM (\a b -> return (f a b))

foldrM :: (AllocM m, Key k, Value a)
       => (a -> b -> m b) -> b -> Tree k a -> m b

by reusing definitions of foldrM for 'Index' and 'Map'?

Comment (skeuchel) [inline src/Data/BTree/Fold.hs]: I just checked the interface of Data.Map. It seems it doesn't export a foldrM, but you can just use Foldable.foldrM for both 'Index' and 'Map'.

File: tests/Properties/Fold.hs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+prop_foldable_toList_fromList :: [(Int64, Integer)] -> Bool
+prop_foldable_toList_fromList kvs
+    | Just l <- runInMemory (createAppendDb "Main"
+                             >>= insertAll kvs
+                             >>= readTransact Tree.toList)
+    = l == F.toList (M.fromList kvs)
+    | otherwise = False
+
+runInMemory :: StoreT String Identity a -> Maybe a
```
Comment (skeuchel) [inline tests/Properties/Fold.hs]: We need infrastructure for testing the impure variant anyway. Maybe you can factor this out into another module in a new PR.

**Comment by hverr** [comment]
@skeuchel see a2e70cbf4fce61d210c43406cbec5d428172ffa2

File: tests/Properties/Fold.hs (hunk 2 above)
Comment (hverr) [inline tests/Properties/Fold.hs]: yes this will probably be dealt with in the future

**Pull request merged**